### PR TITLE
genq: unlimited private queue

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Once the genq private pool can easily shrink, an arbitrary upper limit become no longer critical. We can easily allow certain pools to be unlimited.

[skip warnings]

## Impact

* `rma/manyget` will not fail due to arbitrary genq buffer limit
* ofi am no longer need deferred Q 
  -- looks like the am_isend_pipeline still uses the deferred queue. But at least the code can be greatly simplified.
  -- I think the pipeline can simply be chained through send completion.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
Close in favor of PR #5534 

This PR adds an additional layer of free list mechanism and it is a bit too complex. PR #5534 only frees the slab and thus leaving the meta-layer (blocks and cells management) intact.

**Comment by hzhou** [comment]
Hmm, we really need to clean the freelist as well, otherwise, it will endup keep free and mallocing again the last block. Re-favor this PR over #5534

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

File: src/mpid/common/genq/mpidu_genq_private_pool.c [hunk 1]
```diff
@@ -190,10 +248,23 @@ int MPIDU_genq_private_pool_free_cell(MPIDU_genq_private_pool_t pool, void *cell
...
 
-    cell_h->next = pool_obj->free_list_head;
-    pool_obj->free_list_head = cell_h;
-    cell_h->block->num_used_cells--;
+    block->num_used_cells--;
+
+    if (block->num_used_cells == pool_obj->num_cells_in_block - 1) {
+        append_free_blocks(pool_obj, block);
+    } else if (block->num_used_cells == 0) {
+        if (pool_obj->num_blocks > 1) {
```
Comment (yfguo) [inline src/mpid/common/genq/mpidu_genq_private_pool.c:264]: Can we a comment about the last block is will not be freed until the pool is destroyed?

Comment (yfguo) [reply src/mpid/common/genq/mpidu_genq_private_pool.c:264]: Or making it configurable?

**Review by yfguo — CHANGES_REQUESTED** [review]
LGTM except for a minor comment.

File: src/mpid/common/genq/mpidu_genq_private_pool.c (hunk 1 above)
Comment (hzhou) [reply src/mpid/common/genq/mpidu_genq_private_pool.c:264]: I'll add a comment. Let's avoid the extra configuring complexity until we have a specific reason.

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/gpu/ofi

**Review by yfguo — APPROVED** [review]

**Comment by raffenet** [comment]
> Once the genq private pool can easily shrink, an arbitrary upper limit become no longer critical. We can easily allow certain pools to be unlimited.

FWIW, this was not my understanding for why there was a limit on GenQ buffers. My understanding is that we wanted to set an upper limit on the temporary memory used by MPICH for communication. That is at a certain point, MPICH would stop allocating temporary memory and just work on completing outstanding operations.

**Comment by hzhou** [comment]
> > Once the genq private pool can easily shrink, an arbitrary upper limit become no longer critical. We can easily allow certain pools to be unlimited.
> 
> FWIW, this was not my understanding for why there was a limit on GenQ buffers. My understanding is that we wanted to set an upper limit on the temporary memory used by MPICH for communication. That is at a certain point, MPICH would stop allocating temporary memory and just work on completing outstanding operations.

Do you recall the rationale for setting this upper limit? 

From my point of view, the memory usage including the amount of temporary buffer usage should reflect and adapt to application usage. An arbitrary upper limit regardless of application usage doesn't quite make sense. While the application can assert its control via CVAR, this necessarily makes using our library more complex and it is better to avoid.

**Comment by raffenet** [comment]
Looking at ch3, the SRBuf allocator has no limit, but the RMA window op pool does. The previous buffer pool allocator in ch4 had no limit before it was replaced with GenQ.

So I guess we are not all that consistent. An unlimited private pool was discussed when we were thinking about integrating GenQ into the MPIR-level collectives, but ultimately that didn't go anywhere. Perhaps now is a good time to think about that again?

**Comment by hzhou** [comment]
>  An unlimited private pool was discussed when we were thinking about integrating GenQ into the MPIR-level collectives, but ultimately that didn't go anywhere. Perhaps now is a good time to think about that again?

The challenge in collective temporary buffers is not just the available number of buffers, but more about the unlimited size of the buffers. I guess if we put restrictions of certain collective algorithm -- that they work with message size up to certain limit, then these algorithms will be able to use genq pools. I think this is worth considering for some GPU-specific collective algorithms.

**Pull request merged**